### ISSUE TITLE
LSTMTimestep converges to one value?

### ISSUE THREAD

**Issue by upklio** [BrainJS/brain.js#727]
![Screenshot from 2021-09-02 16-27-39](https://user-images.githubusercontent.com/57267403/131872397-73ff448c-a536-4c78-bd9f-74b8cd728c8a.png)

### Overview
The green line is the latest version of the model, the blue is the previous, this is just my rough testing set up. I am attempting to predict the next 60 candles based off the previous 120 candles. Take note of the arrow markers which show where each `forecast` of predictions start. As you can see, the models fail in doing even a semi-decent job of forecasting. Would expect them to at-least somewhat mimic previous price behaviour. Instead, they almost always converge to the one value or (not shown here) end up in some useless oscillation between two arbitrary values.

### Details
I am unable to get my LSTMTimestep model churning out any decent results after days of trying to fine tune. I am currently using just a small subset of data to try and get this right, `288` ohlc items to be precise, all normalised between `-1/1`, here's a small snippet:

```javascript
const samples = [
  { o: -0.43058824, h: -0.40652557, l: -0.39425418, c: -0.39877685 },
  { o: -0.39890196, h: -0.46443269, l: -0.47443597, c: -0.54116356 },
  { o: -0.54133333, h: -0.52322163, l: -0.49164097, c: -0.48627881 },
]
```

Here is the network and training setup:

```javascript
const batchSize = 288
const netOptions = {
  inputSize: 4,
  hiddenLayers: [8],
  outputSize: 4,
  activation: 'leaky-relu',
}
const trainingOptions = {
  errorThresh: 0.01,
  learningRate: 0.005,
  iterations: 20000,
  log: (details) => console.log(details),
}
const trainingData = _.chunk(samples, batchSize) // Training data looks like: [[...288 objects]]
const net = new brain.recurrent.LSTMTimeStep(netOptions)
const stats = net.train(trainingData, trainingOptions)
```

Take note, I've found myself using all `288` items in a single batch. When experimenting, I noticed that splitting the data up into smaller batches (eg. `5`) resulted in the network not being able to consider values from one batch to another, eg. if the last value of a batch is `0.3` and the first value of the following batch is `0.6`, I train the network and hand it `0.3` to predict the next value and it cannot even predict `0.6`. Confused as to why batch sizes are beneficial at all in trying to forecast price movements?

To forecast, I am simply doing the following:

```javascript
const predictions = net.forecast(chunk, 60)
```

Where `chunk` is just `120` items from the exact same training data. I understand it is good practise to actually split your training data 70/30, 70% to train, 30% to validate. But I don't care about that right now. I just want to see if the net can predict what it's already seen.

So anyway, these predictions are actually happening in a loop, I don't feel I need to share it, but as you can probably tell from my screenshot, after 60 "predictions" have been made, I get the last 120 real ohlc items from where the predictions finished and proceed in predicting the next `60` and so on...

Am I inherently confused on how to use LSTMs? Why am I unable to get any meaningful results? Pointers would really be appreciated. Thanks.

**Comment by upklio** [comment]
![Screenshot from 2021-09-02 17-28-49](https://user-images.githubusercontent.com/57267403/131881882-93fa4b81-80e5-4557-9b46-04e84e7db463.png)

Here's some oscillation examples that can be found when setting batch size of the training data to `5`. Again, completely useless. More variation to start-off with, but then converge to what looks like price oscillation loops.

**Comment by robertleeplummerjr** [comment]
It may be that the data is simply too sophisticated or random and that your net may not be able to learn it.  I would guard against putting too much data in at a time, however, because you net can overfit it.  Meaning it will memorize what you've sent into it.  This is likely what is happening when you put hundreds of values in and the answer I'd like to go with for the time being.

I would probably put a stronger focus on how you net is setup.  Hidden layer size and learning rate will very very important here.
